### PR TITLE
[WIP] Fix signature with underscore variables

### PR DESCRIPTION
**Pull request by eksperimental**
There was a bug that would make latter function definition with _,
loose the definition of a previous underscore variable.

```elixir
  def sample(:ok = _value, _opts),
    do: {:ok, :yes}
  def sample(_, _),
    do: :error
```

This error can be seen in the documentation for `Mix.Archive.Install.check_install_spec/2` generated by ExDoc. second arguments `_opts`  is lost and replaced with `arg2`

### CODE REVIEW

File: lib/elixir/lib/module.ex [hunk 1]
```diff
@@ -673,7 +672,7 @@ defmodule Module do
 
   defp simplify_signature({var, _, atom}, acc) when is_atom(atom) do
     case Atom.to_string(var) do
-      "_" <> rest -> {{String.to_atom(rest), [], Elixir}, acc}
```
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: Would anybody care to explain what is the third element of this tuple?

**Comment by eksperimental** [comment]
I leave it as WIP becauseI still have some questions, and I would like to know if we need to test theses docs functions, something like we would do with ExDoc, or should we be testing this edge case from a test in `Code.get_docs/2`?

File: lib/elixir/lib/module.ex (hunk 1 above)
Comment (josevalim) [inline lib/elixir/lib/module.ex]: The third element of this tuple is the variable context. It doesn't matter in the scenario of pretty printing a variable. We are using it to specify the variable priority (Elixir for low priority, nil for high). See here: https://github.com/eksperimental/elixir/blob/7c4376ee6f2afd6bc76bfcc11e103bc1a2c10af7/lib/elixir/lib/module.ex#L764-L772

So the fix is likely in merge_signatures. You can see the last clause which is where we generate the `arg` one.

**Comment by josevalim** [comment]
I believe the proper fix is to give underscored variables that comes from assignment higher precedence (i.e. store them with `nil` context), because we can assume in this context they are being used for documentation.

Regarding tests, do we have any tests for this functionality? I don't think we need to go through `Code.get_docs(...)` though, we attempt to load the internal docs attribute before.

**Comment by eksperimental** [comment]
so if older signature context is Elixir, and newer signature context nil, which one wins?

**Comment by josevalim** [comment]
`nil`. `nil`always wins.

**Comment by eksperimental** [comment]
@josevalim nope, there are not tests whatsoever that check for the docs. I cannot seem to figure out how to read the internal structure, could you please show me a basic example of how you want to do this. thank you.

**Comment by josevalim** [comment]
@eksperimental I will work on some tests and fix the issue with map_set from the other PR. Than hopefully you can build on top of the fix for this too. :)

**Comment by eksperimental** [comment]
I will update the fix for this issue which also solves map set

**Comment by eksperimental** [comment]
@josevalim, please have a look at the new commit.

**Comment by josevalim** [comment]
@eksperimental I went with a different approach and improved the test suite here: cfe959bc6eb63985c271473077d5b4a6a4d022ca

Hopefully know it will be easier to improve this feature with coverage. Thank you! 
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request closed without merging**